tests/macros/shared/src/main/scala/scala/macros/tests/difference/Difference.scala

```diff
@@ -0,0 +1,26 @@
+package scala.macros.tests.difference
+
+import scala.macros._
+
+case class Diff(entity: Option[String], fields: List[DiffValue])
+
+case class DiffValue(oldValue: Any, newValue: Any)
+
+object Difference {
+  inline def apply[A <: Product](lastState: A, newState: A): Diff = meta {
+    val buff = Term.fresh("buff")
+    val body = A.vals.filter(_.isCase).map { f =>
+      val fnName = f.name.value
+      val v1 = Term.fresh(s"${fnName}1")
+      val v2 = Term.fresh(s"${fnName}2")
+      q"""
+           val $v1 = $lastState.${Term.Name(fnName)}
+           val $v2 = $newState.${Term.Name(fnName)}
+           if($v1 != $v2 ) $buff += DiffValue($v1, $v2)
+        """
+    }
+    q"""val $buff = scala.collection.mutable.ListBuffer[DiffValue]()
```

Comment:
Because someone can theoretically define a variable called `scala`, but a variable called `_root_` is off limits.